core/parentchain/light-client/src/io.rs

```diff
@@ -51,48 +56,28 @@ impl<B: Block, Client: Decode + Encode + Debug> StaticSealedIO for LightClientSe
 	}
 }
 
-pub fn read_or_init_validator<B: Block>(
+pub fn init_validator<B, OCallApi>(
 	params: LightClientInitParams<B::Header>,
-) -> Result<B::Header>
+	ocall_api: Arc<OCallApi>,
+) -> Result<LightValidation<B, OCallApi>>
 where
+	B: Block,
 	NumberFor<B>: finality_grandpa::BlockNumberOps,
+	OCallApi: EnclaveOnChainOCallApi,
 {
-	match params {
-		LightClientInitParams::Grandpa { .. } => {
-			if SgxFile::open(LIGHT_CLIENT_DB).is_err() {
-				info!("[Enclave] ChainRelay DB not found, creating new! {}", LIGHT_CLIENT_DB);
-				return init_validator::<B>(params)
-			}
-
-			let validator =
-				LightClientSeal::<B, GrandpaLightValidation<B>>::unseal_from_static_file()?;
+	let genesis_header = params.get_genesis_header().clone();
+	let authorities = params.get_authorities().cloned().unwrap_or_default();
+	let authority_proof = params.get_authority_proof().cloned().unwrap_or_default();
 
-			let genesis = validator.genesis_hash(validator.num_relays()).unwrap();
-			if genesis == params.get_genesis_header().hash() {
-				info!("Found already initialized light client with Genesis Hash: {:?}", genesis);
-				info!("light client state: {:?}", validator);
-				Ok(validator.latest_finalized_header(validator.num_relays()).unwrap())
-			} else {
-				init_validator::<B>(params)
-			}
-		},
-	}
-}
+	let finality: Arc<Box<dyn Finality<B> + Sync + Send + 'static>> = match params {
+		LightClientInitParams::Grandpa { .. } => Arc::new(Box::new(GrandpaFinality {})),
+		LightClientInitParams::Parachain { .. } => Arc::new(Box::new(ParachainFinality {})),
+	};
 
-fn init_validator<B: Block>(params: LightClientInitParams<B::Header>) -> Result<B::Header>
-where
-	NumberFor<B>: finality_grandpa::BlockNumberOps,
-{
-	match params {
-		LightClientInitParams::Grandpa { authorities, genesis_header, authority_proof } => {
-			let mut validator = GrandpaLightValidation::<B>::new();
+	let mut validator = LightValidation::<B, OCallApi>::new(ocall_api, finality);
 
-			validator.initialize_relay(genesis_header, authorities, authority_proof)?;
-			LightClientSeal::<B, GrandpaLightValidation<B>>::seal_to_static_file(
-				validator.clone(),
-			)?;
+	validator.initialize_relay(genesis_header, authorities, authority_proof)?;
+	LightClientStateSeal::<B, LightValidationState<B>>::seal_to_static_file(validator.get_state())?;
 
```

Comment:
Here is the last tricky part I guess. I think we need to make the relay chain generic over finality related data:

```rust
#[derive(Encode, Decode, Clone, PartialEq)]
pub struct RelayState<Block: BlockT, FinalityData: Decode + Encode> {
	pub last_finalized_block_header: Block::Header,
	pub penultimate_finalized_block_header: Block::Header,
	pub header_hashes: Vec<Block::Hash>,
	pub unjustified_headers: Vec<Block::Hash>, // Finalized headers without grandpa proof
	pub verify_tx_inclusion: Vec<OpaqueExtrinsic>, // Transactions sent by the relay
	pub finality_data: FinalityData
}

struct GrandpaFinalityData<Block: BlockT> {
	pub current_validator_set: AuthorityList,
	pub current_validator_set_id: SetId,
	pub scheduled_change: Option<ScheduledChangeAtBlock<Block::Header>>, // Scheduled Authorities change as indicated in the header's digest.
}
```

Then we can make the initialize relay generic over the finality data:

```rust

impl LightValidation<Block, FinalityData> {
    pub fn initialize_relay(finality_data: FinalityData);
}

```

Reply:
will be done in #776